cli/stork-publisher-agent/websocket.go

```diff
@@ -0,0 +1,421 @@
+package stork_publisher_agent
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/big"
+	"net"
+	"net/http"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gorilla/websocket"
+	"github.com/rs/zerolog"
+)
+
+const FullQueueLogFrequency = time.Second * 10
+const HandshakeTimeout = time.Second * 10
+const ReadBufferSize = 1048576
+const WriteBufferSize = 1048576
+
+type WebsocketConnection struct {
+	conn    *websocket.Conn
+	logger  zerolog.Logger
+	onClose func()
+	closed  chan struct{}
+}
+
+func NewWebsocketConnection(conn *websocket.Conn, logger zerolog.Logger, onClose func()) *WebsocketConnection {
+	return &WebsocketConnection{
+		conn:    conn,
+		logger:  logger,
+		onClose: onClose,
+		closed:  make(chan struct{}),
+	}
+}
+
+// Close notifies the reader and writer threads that the websocket is being closed,
+// then closes the websocket connection and invokes the onClose callback.
+func (ws *WebsocketConnection) Close() {
+	select {
+	case <-ws.closed: // already closed
+	default:
+		close(ws.closed)
+		_ = ws.conn.Close()
+		ws.onClose()
+	}
+}
+
+func (ws *WebsocketConnection) IsClosed() bool {
+	select {
+	case <-ws.closed:
+		return true
+	default:
+		return false
+	}
+}
+
+type IncomingWebsocketConnection struct {
+	WebsocketConnection
+	logger zerolog.Logger
+}
+
+func NewIncomingWebsocketConnection(conn WebsocketConnection, logger zerolog.Logger) *IncomingWebsocketConnection {
+	return &IncomingWebsocketConnection{
+		WebsocketConnection: conn,
+		logger:              logger,
+	}
+}
+
+func convertToValueUpdate(valueUpdatePushWebsocket ValueUpdatePushWebsocket) (*ValueUpdate, error) {
+	var value *big.Float
+	if len(valueUpdatePushWebsocket.ValueString) > 0 {
+		var success bool
+		value, success = new(big.Float).SetString(valueUpdatePushWebsocket.ValueString)
+		if !success {
+			return nil, errors.New("failed to convert string to float")
+		}
+	} else if valueUpdatePushWebsocket.ValueFloat > 0 {
```

Comment:
I want to explicitly fail if a user doesn't pass us either of the value types (rather than just treating it as value zero). This means that we can't actually receive a value of zero though. 

The most correct way to do this would be to include a binary or some kind of "type" indicator to make it clear which field we should look at, but that feels like unnecessary overhead if it's once per message